pkg/ddl/ddl.go

```diff
@@ -792,6 +791,104 @@ func (d *ddl) Start(ctxPool *pools.ResourcePool) error {
 	return nil
 }
 
+// detect versions of all TiDB instances and choose a job version to use, rules:
+//   - if it's in test or run in uni-store, use V2 directly if ForceDDLJobVersionToV1InTest
+//     is not set, else use V1, we use this rule to run unit-tests using V1.
+//   - if all TiDB instances have version >= 8.4.0, use V2
+//   - otherwise, we must during upgrade from lower version, then start a background
+//     routine to detect the version of all TiDB instances repeatedly, when upgrade
+//     is done, we will use V2, and exit the routine.
+//
+// Note: at the time of this PR, some job types hasn't finished migrating to V2,
+// they will stay in V1 regardless of the version we choose here.
+//
+// It's possible that user start a new TiDB of version < 8.4.0 after we detect that
+// all instances have version >= 8.4.0, we will not consider this case here as we
+// don't support downgrade cluster version right now. And even if we try to change
+// the job version in use back to V1, it still will not work when owner transfer
+// to the new TiDB instance which cannot not handle existing submitted jobs of V2.
+func (d *ddl) detectAndUpdateJobVersion() {
+	if d.etcdCli == nil {
+		if ForceDDLJobVersionToV1InTest != "false" {
+			model.SetJobVerInUse(model.JobVersion1)
+			return
+		}
+		model.SetJobVerInUse(model.JobVersion2)
+		return
+	}
+
+	err := d.detectAndUpdateJobVersionOnce()
+	if err != nil {
+		logutil.DDLLogger().Warn("detect job version failed", zap.String("err", err.Error()))
+	}
+
+	if model.GetJobVerInUse() == model.JobVersion2 {
+		return
+	}
+
+	logutil.DDLLogger().Info("job version in use is not v2, maybe in upgrade, start detecting",
+		zap.Stringer("current", model.GetJobVerInUse()))
+	d.wg.RunWithLog(func() {
+		ticker := time.NewTicker(detectJobVerInterval)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+			case <-d.ctx.Done():
+				return
+			}
+			err = d.detectAndUpdateJobVersionOnce()
+			if err != nil {
+				logutil.DDLLogger().Warn("detect job version failed", zap.String("err", err.Error()))
+			}
+			failpoint.InjectCall("afterDetectAndUpdateJobVersionOnce")
+			if model.GetJobVerInUse() == model.JobVersion2 {
+				logutil.DDLLogger().Info("job version in use is v2 now, stop detecting")
+				return
+			}
+		}
+	})
+}
+
+// when all TiDB instances have version >= 8.4.0, we can use job version 2, otherwise
+// we should use job version 1 to keep compatibility.
+func (d *ddl) detectAndUpdateJobVersionOnce() error {
+	infos, err := infosync.GetAllServerInfo(d.ctx)
+	if err != nil {
+		return err
+	}
+	allSupportV2 := true
+	for _, info := range infos {
+		tidbVer := strings.TrimPrefix(info.Version, mysql.ServerVerPrefix)
+		tidbVer = strings.TrimPrefix(tidbVer, "v")
+		ver, err2 := semver.NewVersion(tidbVer)
```

Comment:
In our daily development, `info.Version` may be empty, will this affect our debugging convenience for v2?

```
mysql> select tidb_version();
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                          |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: None
Edition: Community
Git Commit Hash: None
Git Branch: None
UTC Build Time: None
GoVersion: go1.21.0
Race Enabled: false
Check Table Before Drop: false
Store: tikv |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

Reply:
it's really bad, I plan to change the default value of the server-versioon var to `v8.4.0-this-is-a-placeholder`, what do you think.

Reply:
I think it is OK to just leave it at that. A default value needs to be updated for every release?

Reply:
> A default value needs to be updated for every release?

no, it's just a place holder used for debug where tidb-server is built from goland or `go build` directly, otherwise it will have a value from `git describe`